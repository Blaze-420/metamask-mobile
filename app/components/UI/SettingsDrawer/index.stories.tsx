import React from 'react';
import { View } from 'react-native';
import { Meta, StoryObj } from '@storybook/react-native';
import SettingsDrawer from './';
import { IconName } from '../../../component-library/components/Icons/Icon';
import { TextColor } from '../../../component-library/components/Texts/Text';

const backdropStyle = { backgroundColor: 'white', padding: 16 };

const meta: Meta<typeof SettingsDrawer> = {
  title: 'Components/UI/SettingsDrawer',
  component: SettingsDrawer,
  decorators: [(Story) => <View style={backdropStyle}>{Story()}</View>],
};

export default meta;

type Story = StoryObj<React.ComponentProps<typeof SettingsDrawer>>;

const mockPress = () => undefined;

export const Default: Story = {
  args: {
    title: 'Settings Option',
    description: 'This is a description of the settings option',
    onPress: mockPress,
  },
};

export const WithWarning: Story = {
  args: {
    ...Default.args,
    warning: 'Important warning message',
  },
};

export const WithIcon: Story = {
  args: {
    ...Default.args,
    iconName: IconName.Setting,
    iconColor: '#037DD6',
  },
};

export const WithoutArrow: Story = {
  args: {
    ...Default.args,
    renderArrowRight: false,
  },
};

export const CustomTitleColor: Story = {
  args: {
    ...Default.args,
    titleColor: TextColor.Primary,
  },
};

export const FirstItem: Story = {
  args: {
    ...Default.args,
    isFirst: true,
  },
};

export const LastItem: Story = {
  args: {
    ...Default.args,
    isLast: true,
  },
};

export const FirstAndLastItem: Story = {
  args: {
    ...Default.args,
    isFirst: true,
    isLast: true,
  },
};

export const LongContent: Story = {
  args: {
    title:
      'Very Long Settings Option Title That Might Need to Wrap to Multiple Lines',
    description:
      'This is a very long description that contains a lot of text to demonstrate how the component handles long content and wrapping behavior in real-world scenarios.',
    onPress: () => mockPress,
  },
};

export const WithoutDescription: Story = {
  args: {
    title: 'Settings Option',
    onPress: () => mockPress,
  },
};

export const WithWarningAndIcon: Story = {
  args: {
    ...Default.args,
    warning: 'Important warning message',
    iconName: IconName.Danger,
    iconColor: '#D73847',
  },
};
