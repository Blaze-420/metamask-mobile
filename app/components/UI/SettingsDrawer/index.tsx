import React from 'react';
import { View, StyleSheet, TouchableOpacity, Platform } from 'react-native';
import { fontStyles } from '../../../styles/common';
import { useTheme } from '../../../util/theme';
import generateTestId from '../../../../wdio/utils/generateTestId';
import Icon, {
  IconColor,
  IconName,
  IconSize,
} from '../../../component-library/components/Icons/Icon';
import ListItem from '../../../component-library/components/List/ListItem/ListItem';
import ListItemColumn, {
  WidthType,
} from '../../../component-library/components/List/ListItemColumn';
import Text, {
  TextVariant,
  TextColor,
} from '../../../component-library/components/Texts/Text';
import { Colors } from '../../../util/theme/models';
import { SettingsDrawerProps } from './index.types';

const isSettingsRedesignEnabled =
  process.env.MM_SETTINGS_REDESIGN_ENABLE === 'true';

const createStyles = (
  colors: Colors,
  { isFirst, isLast }: { isFirst: boolean; isLast: boolean },
) =>
  StyleSheet.create({
    root: {
      backgroundColor: isSettingsRedesignEnabled
        ? colors.background.alternative
        : colors.background.default,
      padding: 16,
      ...(isFirst && { borderTopLeftRadius: 8, borderTopRightRadius: 8 }),
      ...(!isFirst && {
        borderTopWidth: 1,
        borderTopColor: colors.primary.inverse,
      }),
      ...(isLast && { borderBottomLeftRadius: 8, borderBottomRightRadius: 8 }),
    },
    settingsCard: {
      alignItems: 'center',
      backgroundColor: colors.background.alternative,
      padding: 16,
    },
    action: {
      paddingLeft: 16,
    },
    warningTag: {
      flexDirection: 'row',
      alignSelf: 'flex-start',
      alignItems: 'center',
      height: 24,
      paddingHorizontal: 8,
      marginTop: 8,
      borderRadius: 12,
      backgroundColor: colors.error.muted,
    },
    warningText: {
      marginLeft: 4,
    },
    menuItemWarningText: {
      color: colors.text.default,
      fontSize: 12,
      ...fontStyles.normal,
    },
    icon: {
      marginRight: 12,
    },
    separator: {
      borderTopWidth: 1,
      borderTopColor: colors.primary.inverse,
    },
  });

const SettingsDrawer: React.FC<SettingsDrawerProps> = ({
  title,
  description,
  onPress,
  iconName,
  iconColor,
  warning,
  renderArrowRight = true,
  isFirst = false,
  isLast = false,
  testID,
  titleColor = TextColor.Default,
}) => {
  const { colors } = useTheme();
  const styles = createStyles(colors, { isFirst, isLast });
  const titleVariant = isSettingsRedesignEnabled
    ? TextVariant.BodyMDMedium
    : TextVariant.BodyLGMedium;
  const descriptionVariant = isSettingsRedesignEnabled
    ? TextVariant.BodySM
    : TextVariant.BodyMD;

  return (
    <TouchableOpacity onPress={onPress} {...generateTestId(Platform, testID)}>
      <ListItem style={styles.root} gap={16}>
        {iconName && isSettingsRedesignEnabled && (
          <Icon
            name={iconName}
            size={IconSize.Md}
            color={iconColor || colors.text.default}
          />
        )}
        <ListItemColumn widthType={WidthType.Fill}>
          <Text variant={titleVariant} color={titleColor}>
            {title}
          </Text>
          {description && (
            <Text variant={descriptionVariant} color={TextColor.Alternative}>
              {description}
            </Text>
          )}
          {warning && (
            <View style={styles.warningTag}>
              <Icon
                size={IconSize.Sm}
                color={IconColor.Error}
                name={IconName.Danger}
              />
              <Text
                variant={TextVariant.BodyMD}
                color={TextColor.Error}
                style={styles.warningText}
              >
                {warning}
              </Text>
            </View>
          )}
        </ListItemColumn>
        {renderArrowRight && (
          <ListItemColumn>
            <Icon
              style={styles.action}
              size={IconSize.Md}
              name={IconName.ArrowRight}
              color={IconColor.Default}
            />
          </ListItemColumn>
        )}
      </ListItem>
    </TouchableOpacity>
  );
};

export default SettingsDrawer;
