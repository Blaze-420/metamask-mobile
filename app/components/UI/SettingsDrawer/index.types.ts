import { IconName } from '../../../component-library/components/Icons/Icon';
import { TextColor } from '../../../component-library/components/Texts/Text';

export interface SettingsDrawerProps {
  title: string;
  /**
   * Additional descriptive text about this option
   */
  description?: string;
  /**
   * Disable bottom border
   */
  noBorder?: boolean;
  /**
   * Handler called when this drawer is pressed
   */
  onPress?: () => void;
  /**
   * Display SettingsNotification
   */
  warning?: string;
  /**
   * Icon name
   */
  iconName?: IconName;
  /**
   * Icon color
   */
  iconColor?: string;
  /**
   * Display arrow right
   */
  renderArrowRight?: boolean;
  /**
   * First item
   */
  isFirst?: boolean;
  /**
   * Last item
   */
  isLast?: boolean;
  /**
   * Test id for testing purposes
   */
  testID?: string;
  /**
   * Title color
   */
  titleColor?: TextColor;
}
