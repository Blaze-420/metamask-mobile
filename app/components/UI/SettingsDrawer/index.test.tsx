import React from 'react';
import { render } from '@testing-library/react-native';
import SettingsDrawer from './';
import { IconName } from '../../../component-library/components/Icons/Icon';

const originalEnv = process.env;

describe('SettingsDrawer', () => {
  const originalProps = {
    title: 'Test Title',
    onPress: jest.fn(),
    description: 'Test Description',
    renderArrowRight: true,
  };
  const redesignProps = {
    ...originalProps,
    iconName: IconName.Setting,
    iconColor: 'red',
  };

  beforeEach(() => {
    jest.clearAllMocks();
    process.env = { ...originalEnv };
  });

  afterAll(() => {
    process.env = originalEnv;
  });

  describe('SettingsDrawer', () => {
    it('should render correctly', () => {
      process.env.MM_SETTINGS_REDESIGN_ENABLE = 'false';
      const { toJSON } = render(<SettingsDrawer {...originalProps} />);
      expect(toJSON()).toMatchSnapshot();
    });
    it('should render with redesign enabled', () => {
      process.env.MM_SETTINGS_REDESIGN_ENABLE = 'true';
      const { toJSON } = render(<SettingsDrawer {...redesignProps} />);
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
