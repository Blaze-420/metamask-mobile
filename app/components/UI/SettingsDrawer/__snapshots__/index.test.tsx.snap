// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsDrawer SettingsDrawer should render correctly 1`] = `
<TouchableOpacity
  onPress={[MockFunction]}
>
  <View
    accessibilityRole="none"
    accessible={true}
    style={
      {
        "backgroundColor": "#ffffff",
        "borderTopColor": "#ffffff",
        "borderTopWidth": 1,
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
        testID="listitemcolumn"
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#141618",
              "fontFamily": "EuclidCircularB-Medium",
              "fontSize": 16,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Test Title
        </Text>
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#6a737d",
              "fontFamily": "EuclidCircularB-Regular",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 22,
            }
          }
        >
          Test Description
        </Text>
      </View>
      <View
        accessible={false}
        style={
          {
            "width": 16,
          }
        }
        testID="listitem-gap"
      />
      <View
        style={
          {
            "flex": -1,
          }
        }
        testID="listitemcolumn"
      >
        <SvgMock
          color="#141618"
          height={20}
          name="ArrowRight"
          style={
            {
              "height": 20,
              "paddingLeft": 16,
              "width": 20,
            }
          }
          width={20}
        />
      </View>
    </View>
  </View>
</TouchableOpacity>
`;

exports[`SettingsDrawer SettingsDrawer should render with redesign enabled 1`] = `
<TouchableOpacity
  onPress={[MockFunction]}
>
  <View
    accessibilityRole="none"
    accessible={true}
    style={
      {
        "backgroundColor": "#ffffff",
        "borderTopColor": "#ffffff",
        "borderTopWidth": 1,
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
        testID="listitemcolumn"
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#141618",
              "fontFamily": "EuclidCircularB-Medium",
              "fontSize": 16,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Test Title
        </Text>
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#6a737d",
              "fontFamily": "EuclidCircularB-Regular",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 22,
            }
          }
        >
          Test Description
        </Text>
      </View>
      <View
        accessible={false}
        style={
          {
            "width": 16,
          }
        }
        testID="listitem-gap"
      />
      <View
        style={
          {
            "flex": -1,
          }
        }
        testID="listitemcolumn"
      >
        <SvgMock
          color="red"
          height={20}
          name="ArrowRight"
          style={
            {
              "height": 20,
              "paddingLeft": 16,
              "width": 20,
            }
          }
          width={20}
        />
      </View>
    </View>
  </View>
</TouchableOpacity>
`;
